bootstore/src/schemes/v0/state.rs

```diff
@@ -0,0 +1,594 @@
+        match self {
+            RackSecretState::Retrieving { expiry, .. } => {
+                if *expiry < now {
+                    *self = RackSecretState::Empty;
+                    return ApiError::RackSecretLoadTimeout.into();
+                }
+            }
+            RackSecretState::Computed { expiry, .. } => {
+                if *expiry < now {
+                    // This is just cleanup of the secret from memory
+                    // No reason to report this
+                    *self = RackSecretState::Empty;
+                }
+            }
+            RackSecretState::Empty => (),
+        }
+        Output::none()
+    }
+
+    fn validate_share(
+        from: &Baseboard,
+        share: &Vec<u8>,
+        share_digests: &Vec<Sha3_256Digest>,
+    ) -> Result<(), ApiError> {
+        let computed_hash = Sha3_256Digest(
+            Sha3_256::digest(share).as_slice().try_into().unwrap(),
+        );
+
+        if !share_digests.contains(&computed_hash) {
+            Err(ApiError::InvalidShare { from: from.clone() })
+        } else {
+            Ok(())
+        }
+    }
+
+    // Send a request for a share to each peer
+    pub fn broadcast_share_requests(
+        &self,
+        rack_uuid: Uuid,
+        peers: &BTreeSet<Baseboard>,
+    ) -> (Uuid, Output) {
+        // All broadcast requests share the same id
+        let request_id = Uuid::new_v4();
+        let envelopes = peers
+            .iter()
+            .cloned()
+            .map(|to| Envelope {
+                to,
+                msg: Request {
+                    id: request_id,
+                    type_: RequestType::GetShare { rack_uuid },
+                }
+                .into(),
+            })
+            .collect();
+
+        (request_id, Output { persist: false, envelopes, api_output: None })
+    }
+
+    /// We received a share
+    ///
+    /// Verify that the share is valid and recompute the secret if there are
+    /// enough shares.
+    pub fn on_share(
+        &mut self,
+        from: Baseboard,
+        request_id: Uuid,
+        share: Vec<u8>,
+        rack_secret_expiry: Ticks,
+    ) -> Output {
+        match self {
+            RackSecretState::Empty => ApiError::UnexpectedResponse {
+                from,
+                state: "-",
+                request_id,
+                msg: "share",
+            }
+            .into(),
+            RackSecretState::Retrieving {
+                from: from_all,
+                shares,
+                threshold,
+                share_digests,
+                ..
+            } => {
+                // Ignore the share if we already have one from this peer
+                if from_all.contains(&from) {
+                    return Output::none();
+                }
+
+                // Compute and validate hash of the received key share
+                if let Err(api_error) =
+                    Self::validate_share(&from, &share, &share_digests)
```

Comment:
It's too bad the return type isn't a `Result` here, so we can't do
```rust
Self::validate_share(&from, &share, &share_digests)?;
```

Reply:
Fixed in [98ada27](https://github.com/oxidecomputer/omicron/pull/3385/commits/98ada27b53d4429463cde94cc91bedec0bc4ef98)